modules/xcm-bridge-hub-router/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+/// The factor that is used to increase current message fee factor for every sent kilobyte.
+const MESSAGE_SIZE_FEE_BASE: FixedU128 = FixedU128::from_rational(1, 1000); // 0.001
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::config]
+	pub trait Config<I: 'static = ()>: frame_system::Config {
+		/// Universal location of this runtime.
+		type UniversalLocation: Get<InteriorMultiLocation>;
+		/// Relative location of the sibling bridge hub.
+		type SiblingBridgeHubLocation: Get<MultiLocation>;
+		/// The bridged network that this config is for.
+		type BridgedNetworkId: Get<NetworkId>;
+
+		/// Actual message sender (XCMP/DMP) to the sibling bridge hub location.
+		type ToBridgeHubSender: SendXcm + LocalXcmQueue;
+
+		/// Base bridge fee that is paid for every outbound message.
+		type BaseFee: Get<u128>;
+		/// Additional fee that is paid for every byte of the outbound message.
+		type ByteFee: Get<u128>;
+		/// Asset that is used to paid bridge fee.
+		type FeeAsset: Get<AssetId>;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
+
+	#[pallet::hooks]
+	impl<T: Config<I>, I: 'static> Hooks<BlockNumberFor<T>> for Pallet<T, I> {
+		fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
+			// if XCM queue is still overloaded, we don't change anything
+			if T::ToBridgeHubSender::is_overloaded() {
+				return Weight::zero() // TODO: benchmarks
+			}
+
+			DeliveryFeeFactor::<T, I>::mutate(|f| {
+				*f = InitialFactor::get().max(*f / EXPONENTIAL_FEE_BASE);
+				*f
+			});
+
+			Weight::zero() // TODO: benchmarks
+		}
+	}
+
+	/// Initialization value for the delivery fee factor.
+	#[pallet::type_value]
+	pub fn InitialFactor() -> FixedU128 {
+		FixedU128::one()
+	}
+
+	/// The number to multiply the base delivery fee by.
+	#[pallet::storage]
+	#[pallet::getter(fn delivery_fee_factor)]
+	pub type DeliveryFeeFactor<T: Config<I>, I: 'static = ()> =
+		StorageValue<_, FixedU128, ValueQuery, InitialFactor>;
+
+	impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		/// Called when new message is sent (queued to local outbound XCM queue) over the bridge.
+		pub(crate) fn on_message_sent_to_bridge(message_size: u32) {
+			// if outbound queue is not overloaded, do nothing
+			if !T::ToBridgeHubSender::is_overloaded() {
+				return
+			}
+
+			// ok - we need to increase the fee factor, let's do that
+			let message_size_factor = FixedU128::from_u32(message_size.saturating_div(1024))
+				.saturating_mul(MESSAGE_SIZE_FEE_BASE);
+			let total_factor = EXPONENTIAL_FEE_BASE.saturating_add(message_size_factor);
+			DeliveryFeeFactor::<T, I>::mutate(|f| {
+				*f = f.saturating_mul(total_factor);
+				*f
+			});
+		}
+	}
+}
+
+/// We'll be using `SovereignPaidRemoteExporter` to send remote messages over the sibling/child
+/// bridge hub.
+type ViaBridgeHubExporter<T, I> = SovereignPaidRemoteExporter<
```

Comment:
Two things here:
1) we need to check that the bridge hub doesn't allow unpaid execution after this is changed;
2) in v2 - do we need to check at the bridge hub that the fee is enough? And how to do that?

Reply:
Actually both points may be done for v2 - in v1 we only deal with system parachains. But for v2 we need to check that the execution is paid.

As for the 2nd - IIUC we don't need to check anything apart from the fact that the execution is paid. If the sending chain is not able to limit the messages flow, it'll end up with messages piling up at its storage. And checking fee value here (at the source bridge chain) doesn't matter much.